### PR TITLE
integration runner refactor + sharded dry-runs

### PR DESCRIPTION
**Pull request by geoberle**
This PR refactors the current `cli.py` based integration startup function and enhances it with `sharded` dry-runs.

REVIEW COMMIT BY COMMIT

# Part 1 - Refactor
commit 1

The main logic for integration running has been moved to the `reconcile.utils.runtime` module. An integration and the required callback hooks `run` and `get_early_exit_desired_state` are now part of a class `QontractReconcileIntegration`. New integrations can subclass this one directly and implement the hook methods. Existing integrations are wrapped during startup into an `ModuleBasedQontractReconcileIntegration`. This way the code in `reconcile.utils.runtime.runner` deal only with `QontractReconcileIntegration` and can rely on a typed interface.

And integration can be executed by constructing a `IntegrationRunConfiguration` object which contains all the relevant generic qontract-reconcile runtime infos (`dry_run`, `validate_schema`, ...)

# Part 2 - Detect affected shards
commit 2+3

the difference in desired state state is used to find to find the affected shards for an integration. an integration can use the `get_desired_state_shard_config` function to return a `DesiredStateShardConfig` object that defines where to find the affected shards in the desired state.

the `DesiredStateShardConfig` object can contain a callback function where the integration can decide, if sharded run should be tried based on the detected shards, e.g. this can be used to not run in sharded mode if too many shards have been detected.

Sharded dry-runs are opt-int in many levels: 
* an integration must provide early-exit functionality and a `DesiredStateShardConfig` configuration
* an integration must be started with the `--check-only-affected-shards` CLI flag. This flag can be controlled via app-interface hack scripts (through data in the `/app-sre/integration-1.yml` [schema](https://github.com/app-sre/qontract-schemas/pull/373)) to activate/deactivate the functionality on demand

# Part 3 - Enable `terraform-resources` for sharded dry-runs

(!) this part will be split out into a dedicated PR - i left it here for now to show how the `desired_state_shard_config` hook works.

by implementing the `desired_state_shard_config` function, the `terraform-resources` integrations opts into the sharded integration dry-run mechanism.

the desired state returned by `early_exit_desired_state` has been enhanced by adding `IDENTIFIER_FIELD_NAME` fields to establish identifies for objects in lists, which speeds up the difference finding procedure. `IDENTIFIER_FIELD_NAME` is not something new but is populated in `bundles` by `qontract-validator` when a schema defines `uniqe` or `context-unique` fields.

this is not really necessary but helps in speeding up the diffing process so it does not run into timeouts, which would prevent sharded dry-runs.

Jira: https://issues.redhat.com/browse/APPSRE-6785
Schema change for activation flag: https://github.com/app-sre/qontract-schemas/pull/373

Signed-off-by: Gerd Oberlechner <goberlec@redhat.com>

### CODE REVIEW

File: reconcile/cli.py
Comment (chassing) [inline reconcile/cli.py:442]: Please raise a dedicated exception (e.g., `UnknownIntegrationTypeError`) instead of `Exception`

File: reconcile/test/runtime/fixtures.py
Comment (chassing) [inline reconcile/test/runtime/fixtures.py:12]: Please don't name it `Test...` otherwise pytest thinks it's testcase and shows this warning

```
11:37:18 reconcile/test/runtime/fixtures.py:12
11:37:18   /package/reconcile/test/runtime/fixtures.py:12: PytestCollectionWarning: cannot collect test class 'TestIntegration' because it has a __init__ constructor (from: reconcile/test/runtime/test_utils_runtime_runner.py)
11:37:18     class TestIntegration(QontractReconcileIntegration):
```

File: reconcile/test/runtime/fixtures.py
Comment (chassing) [inline reconcile/test/runtime/fixtures.py:1]: Maybe a personal taste, but I would use a `conftest.py` instead of this file and use the [pytest fixtures](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html) feature to consume `TestIntegration` or `ShardableTestIntegration` instances in my test cases

File: reconcile/utils/runtime/desired_state_diff.py
Comment (chassing) [inline reconcile/utils/runtime/desired_state_diff.py:79]: A `__doc__` would be fine.

File: reconcile/utils/runtime/desired_state_diff.py
Comment (chassing) [inline reconcile/utils/runtime/desired_state_diff.py:82]: A `__doc__` would be fine.

File: reconcile/utils/runtime/desired_state_diff.py
Comment (chassing) [inline reconcile/utils/runtime/desired_state_diff.py:114]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:15]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:21]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:27]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:35]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:39]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:43]: A `__doc__` would be fine.

File: reconcile/utils/runtime/integration.py
Comment (chassing) [inline reconcile/utils/runtime/integration.py:56]: A `__doc__` would be fine.

File: reconcile/utils/runtime/runner.py
Comment (chassing) [inline reconcile/utils/runtime/runner.py:78]: A `__doc__` would be fine.

File: reconcile/utils/runtime/runner.py
Comment (chassing) [inline reconcile/utils/runtime/runner.py:111]: A `__doc__` would be fine.

**Review by chassing — COMMENTED** [review]
Some nits :)

File: reconcile/test/runtime/fixtures.py
Comment (geoberle) [reply reconcile/test/runtime/fixtures.py]: i see. makes sense. renamed in 0c630fb625ff1f4d83889ba63e0df7f0fb4466fb

File: reconcile/test/runtime/fixtures.py
Comment (geoberle) [reply reconcile/test/runtime/fixtures.py:1]: fixed in e6e3d6453535b1f29bc7615f1ee065c45fa06c64

File: reconcile/test/runtime/test_utils_runtime_runner.py
Comment (geoberle) [reply reconcile/test/runtime/test_utils_runtime_runner.py]: naming suggestions highly welcome :D

File: reconcile/utils/runtime/runner.py
Comment (geoberle) [reply reconcile/utils/runtime/runner.py:111]: done

File: reconcile/utils/runtime/runner.py
Comment (geoberle) [reply reconcile/utils/runtime/runner.py:69]: done

File: reconcile/utils/runtime/integration.py
Comment (geoberle) [reply reconcile/utils/runtime/integration.py:86]: done

File: reconcile/utils/runtime/integration.py
Comment (geoberle) [reply reconcile/utils/runtime/integration.py]: done

File: reconcile/utils/runtime/integration.py
Comment (geoberle) [reply reconcile/utils/runtime/integration.py]: done

File: reconcile/cli.py
Comment (geoberle) [reply reconcile/cli.py]: done

**Review by maorfr — APPROVED** [review]
:rubber-stamp:

**Review by chassing — APPROVED** [review]
LGTM

File: reconcile/utils/runtime/desired_state_diff.py
Comment (steveteahan) [inline reconcile/utils/runtime/desired_state_diff.py:42]: Reminder on abstracts over concrete inputs (concrete outputs are fine).

`dict` -> `Mapping`
`list` -> `Iterable`

and all that good stuff unless there's a good reason not to.

File: reconcile/utils/runtime/integration.py
Comment (steveteahan) [inline reconcile/utils/runtime/integration.py:38]: +1 to making this a class and `ABC`. Definitely time to start using this pattern moving forward.

File: reconcile/utils/runtime/runner.py
Comment (steveteahan) [inline reconcile/utils/runtime/runner.py:20]: Docstring cover purpose and what each member is used for?

Our CI isn't enforcing this, so I'm not strongly opinionated, but a standard way we can doc these parameters to make them distinct from general comments: https://peps.python.org/pep-0224/

**Review by steveteahan — COMMENTED** [review]
I did a very quick review, but the concept makes sense at a high-level and I like the approach of adding an abstract class to make this more uniform. I think that adding more docstrings and explanations for class members will make it a lot easier for the next person who has to modify the code (although it looks like the test coverage is good and well-written). This is only for public classes/functions and things that wouldn't be very obvious to someone with not a lot of previous knowledge.

No need to block on my approval. If I get time to spend more time on it before you merge I'll provide any other feedback.

File: reconcile/utils/runtime/runner.py
Comment (geoberle) [reply reconcile/utils/runtime/runner.py:20]: I was not aware of attribute docstrings. TIL. I will add respective descriptions.

**Comment by lechuk47** [comment]
I'm chewing this one. Do you have an example of how `DesiredStateConfig` would look to identify shards in integrations like `openshift-resources` where there is no such direct split as in `terraform-resources`?

**Comment by geoberle** [comment]
> I'm chewing this one. Do you have an example of how `DesiredStateConfig` would look to identify shards in integrations like `openshift-resources` where there is no such direct split like in `terraform-resources`?

as you said, for `tf-resources`, there is a natural split of data based on aws-accounts. for `openshift-resources` where there is no balanced natural split, you need to find an artificial one. consider the following approach:

the base data that is looked at to find the affected shards is the data returned from `early_exit_desired_state`. you can change this function to add the artificial shard information to that data. e.g. `$cluster-$first_letter_for_namespace` (this is not really a proposal, just something simple to get the idea across). 

```json
{
    "namespaces":[
      {
         "shard": "cluster1-n",  <-- construct this in code in early_exit_desired_state
         "name": "my-namespace",
         ... regular namespace data ...
      }
    ],
    "resources": [
       <-- something similar here
    ],
}
```

the integration run function needs a `kwarg` to receive such a shard, e.g. `shard`. the integration then needs to process only those resources that belong to the affected artificial shard by applying the same consistent approach to resource selection that was used to build the artificial shard in `early_exit_desired_state`

then the `DesiredStateConfig` object would look like this

```python
DesiredStateShardConfig(
        shard_arg_name="shard",
        shard_path_selectors={
            "namespaces[*].shard",
            "resources[*].wherever.the.shard.can.be.found.here",
        },
        sharded_run_review=lambda proposal: len(proposal.proposed_shards) ==1, # only run sharded when 1 shard is affected
 )
```

this is just a rough sketch how it could be done with artificially constructed shards.

any thoughts?

**Pull request merged**
